tools/src/git_ops.rs

```diff
@@ -43,3 +43,164 @@ pub fn get_blame_lines(git_data: Option<&GitData>, blame_commit: &Option<Commit>
+            // shift remains the same as before
+            continue;
+        }
+        if hunk.new_lines() > hunk.old_lines() {
+            // Lines were added in the hunk, let's map all the new
+            // lines to the last line of the hunk in the old file.
+            let extra_lines = hunk.new_lines() - hunk.old_lines();
+            for i in 0..extra_lines {
+                shift -= 1;
+                adjustments.push((hunk.new_start() + hunk.old_lines() + i, shift))
+            }
+        }
+        if hunk.new_lines() < hunk.old_lines() {
+            // Lines were removed in the hunk, so we have a discontinuity
+            // where lines after the hunk in the new file have a bigger
+            // shift than lines before/inside the hunk.
+            shift = shift + hunk.old_lines() as i32 - hunk.new_lines() as i32;
+            adjustments.push((hunk.new_start() + hunk.new_lines(), shift));
+        }
+    }
+    Ok(LineMap{ adjustments })
+}
+
+/// Given a GitData, commit revision, and target file, this function returns information
+/// to track lines in that file backwards by one commit. Specifically, it returns
+/// the parent commit, the corresponding path of the equivalent source file in that commit
+/// (in case it was moved or copied), and a LineMap that allows mapping individual lines
+/// backwards from the target file to the source file. Note that this only works with
+/// non-merge revisions (i.e. there has to be a unique parent).
+/// If the target_file was not modified in the indicated commit, this function will return
+/// an error.
+fn map_to_previous_version(
+        git_data: &GitData,
+        rev: &str,
+        target_file: &Path
+) -> Result<(Oid, PathBuf, LineMap), Error> {
+    // TODO: cache the result of this function as it will likely get called many times
+    // with the exact same arguments repeatedly
+
+    let commit_obj = git_data.repo.revparse_single(rev)?;
+    let commit = commit_obj.as_commit().ok_or_else(|| Error::from_str("Commit_obj error"))?;
+    if commit.parent_ids().len() != 1 {
+        // If the commit didn't have a unique parent, let's abort
+        return Err(Error::from_str("No unique parent, don't know where to look for prev blame"));
+    }
+
+    let parent_commit = commit.parent(0)?;
+    let older_tree = parent_commit.tree()?;
+    let newer_tree = commit.tree()?;
+
+    let mut diff = git_data.repo.diff_tree_to_tree(Some(&older_tree), Some(&newer_tree), None)?;
+    diff.find_similar(Some(DiffFindOptions::new().renames(true)))?;
+
+    // There should be exactly one delta with the target_file as the "new file"
+    let mut deltas = diff.deltas().filter(|delta| delta.new_file().path() == Some(target_file));
+    let delta = match deltas.next() {
+        Some(d) => {
+            if deltas.next().is_some() {
+                warn!("Found extra delta entries with target {:?} in rev {}", target_file, rev);
+                return Err(Error::from_str("Too many deltas"));
```

Comment:
I use a warn here because if we run into this problem it's a bad assumption in my code. Recall that `warn!` output will end up in an email sent to the searchfox mailing list, so it's magic that way. But yeah I `info!` the `Err` description as you noticed.